[Ingest Pipelines] Error messages

jloleysens opened the pull request:
## Summary

Compresses error feedback when attempting to create or update a pipeline.

Currently pipeline errors are reported as one continuous string (see screenshot). This PR adds code for parsing and displaying errors in a more user friendly way.

## How to test

1. Start Kibana on a basic license
2. Go the ingest pipelines app in the management section
3. Create a new pipeline, adding a processor with an empty `{}` for config (so you know it is invalid).
4. Duplicate the above processor a 6 times
5. Create the pipeline and the error messages should be present at the top of the creation form along with a toast of the error.

## Additional notes

Replicates the pattern in Snapshot & Restore of collapsible indices list

## Screenshots

**Current errors**

<img width="1184" alt="Screenshot 2020-06-18 at 11 59 00" src="https://user-images.githubusercontent.com/8155004/85987021-5ef93700-b9ed-11ea-906e-ac446803cdb7.png">

**After changes**

<img width="1449" alt="Screenshot 2020-06-30 at 11 53 33" src="https://user-images.githubusercontent.com/8155004/86112551-66414300-bac8-11ea-965b-27f13ff53141.png">

<img width="1262" alt="Screenshot 2020-06-30 at 11 53 53" src="https://user-images.githubusercontent.com/8155004/86112557-693c3380-bac8-11ea-879f-9f52d8fb425e.png">

jloleysens commented:
@elasticmachine merge upstream

elasticmachine commented:
Pinging @elastic/es-ui (Team:Elasticsearch UI)

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57046/)
* Commit: bc6cc68b530a860be716d4f7cc7e819c9d24d3b8

### Build metrics
> ✅ unchanged

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57046/","number":"57046","commit":"bc6cc68b530a860be716d4f7cc7e819c9d24d3b8"}]}
PIPELINE-->

cjcenizal commented:
@jloleysens Could you add a screenshot of how the error looks with these changes?

jloleysens commented:
@elasticmachine merge upstream

jloleysens commented:
@cjcenizal Made some minor adjustments to copy and layout and updated the PR description with screenshots.

mdefazio commented:
This is much better! Does the 'Show 3 errors' need to read 'Show 3 more errors'? or '+ 3 more'?

Can we limit the number shown by default (before hiding them) down to 4 or 5 or so?

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57442/)
* Commit: 857360656bf1bb69c66e815a4b67d33c0e6a3864

### Build metrics
> ✅ unchanged

### History
* :green_heart: [Build #57046](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57046/) succeeded bc6cc68b530a860be716d4f7cc7e819c9d24d3b8

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57046/","number":"57046","commit":"bc6cc68b530a860be716d4f7cc7e819c9d24d3b8"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57442/","number":"57442","commit":"857360656bf1bb69c66e815a4b67d33c0e6a3864"}]}
PIPELINE-->

jloleysens commented:
@mdefazio I took the copy from the collapsible list we have in snapshot & restore which does not have the "more" and "less" in it. We can ask someone from docs to also weigh in.

Currently the number of errors is limited to 5, we can go down to 4 if you think that would look better?

mdefazio commented:
> @mdefazio I took the copy from the collapsible list we have in snapshot & restore which does not have the "more" and "less" in it. We can ask someone from docs to also weigh in.
> 
> Currently the number of errors is limited to 5, we can go down to 4 if you think that would look better?

Ah sorry, I mixed up the screenshots trying to go too fast.  5 is good.  :D

My initial reaction was that it wasn't clear enough that hitting the link would show 3 additional errors.  If there are 5 shown, and 5 hidden, maybe this would be confusing. I think it's good we stay consistent, but might be worth having someone from docs weigh in.

alisonelizabeth commented on x-pack/plugins/ingest_pipelines/public/application/components/pipeline_form/pipeline_form_error.tsx:
```diff
@@ -4,29 +4,161 @@
+  const { services } = useKibana();
+  const [isShowingAllErrors, setIsShowingAllErrors] = useState<boolean>(false);
+  const safeErrorResult = toKnownError(error);
+  const hasMoreErrors = safeErrorResult.errors.length > 5;
```
What do you think about extracting `5` as a const, so it's clear it represents the initial number of errors that will be shown?

jloleysens commented:
@elasticmachine merge upstream

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57832/)
* Commit: e6bd1ae327febe57806ff32d72c4b3f385f4aea9
* [Pipeline Steps](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57832/flowGraphTable) (look for red circles / failed steps)
* [Interpreting CI Failures](https://www.elastic.co/guide/en/kibana/current/interpreting-ci-failures.html)

### Failed CI Steps
* [Execute kibana-intake](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57832/execution/node/320/log)

### Build metrics
> ✅ unchanged

### History
* :green_heart: [Build #57442](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57442/) succeeded 857360656bf1bb69c66e815a4b67d33c0e6a3864
* :green_heart: [Build #57046](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57046/) succeeded bc6cc68b530a860be716d4f7cc7e819c9d24d3b8

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57046/","number":"57046","commit":"bc6cc68b530a860be716d4f7cc7e819c9d24d3b8"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57442/","number":"57442","commit":"857360656bf1bb69c66e815a4b67d33c0e6a3864"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57832/","number":"57832","commit":"e6bd1ae327febe57806ff32d72c4b3f385f4aea9"}]}
PIPELINE-->

jloleysens commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57863/)
* Commit: 3176d9b20cb8e7ead33abaa5cc9bc7dc8769e860

### Build metrics
> ✅ unchanged

### History
* :broken_heart: [Build #57832](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57832/) failed e6bd1ae327febe57806ff32d72c4b3f385f4aea9
* :green_heart: [Build #57442](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57442/) succeeded 857360656bf1bb69c66e815a4b67d33c0e6a3864
* :green_heart: [Build #57046](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57046/) succeeded bc6cc68b530a860be716d4f7cc7e819c9d24d3b8

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57046/","number":"57046","commit":"bc6cc68b530a860be716d4f7cc7e819c9d24d3b8"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57442/","number":"57442","commit":"857360656bf1bb69c66e815a4b67d33c0e6a3864"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57832/","number":"57832","commit":"e6bd1ae327febe57806ff32d72c4b3f385f4aea9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57863/","number":"57863","commit":"3176d9b20cb8e7ead33abaa5cc9bc7dc8769e860"}]}
PIPELINE-->

jloleysens commented:
@alisonelizabeth @mdefazio thanks both for the feedback.

I believe I have addressed the feedback points provided. Would you please take another look?

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57942/)
* Commit: b41e08f4ade0d3b90d191b4fa7c956b3b84c63bf

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/b41e08f4ade0d3b90d191b4fa7c956b3b84c63bf) | diff | [baseline](https://github.com/elastic/kibana/commit/518e88cf287ee8fdfefc51ae5040765371d8065b) |
| --- | --- | --- | --- |
| ingestPipelines | 245 | +5 | 240 |

### History
* :green_heart: [Build #57863](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57863/) succeeded 3176d9b20cb8e7ead33abaa5cc9bc7dc8769e860
* :broken_heart: [Build #57832](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57832/) failed e6bd1ae327febe57806ff32d72c4b3f385f4aea9
* :green_heart: [Build #57442](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57442/) succeeded 857360656bf1bb69c66e815a4b67d33c0e6a3864
* :green_heart: [Build #57046](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57046/) succeeded bc6cc68b530a860be716d4f7cc7e819c9d24d3b8

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57046/","number":"57046","commit":"bc6cc68b530a860be716d4f7cc7e819c9d24d3b8"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57442/","number":"57442","commit":"857360656bf1bb69c66e815a4b67d33c0e6a3864"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57832/","number":"57832","commit":"e6bd1ae327febe57806ff32d72c4b3f385f4aea9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57863/","number":"57863","commit":"3176d9b20cb8e7ead33abaa5cc9bc7dc8769e860"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57942/","number":"57942","commit":"b41e08f4ade0d3b90d191b4fa7c956b3b84c63bf"}]}
PIPELINE-->
